### ISSUE TITLE
Feature Request: Battery Charging status

### ISSUE THREAD

**Issue by m14t** [owntracks/talk#32]
Hello,

Firstly, thank you for such a great app, and all the work that you do.  I truly appreciate it.

In the [MQTT JSON body](http://owntracks.org/booklet/tech/json/#_typelocation), I noticed that I have access to:
> * `batt` Device battery level (_iOS,Android/integer/percent/optional_)

In addition to this, it would be great if we could know if the phone was currently plugged in or not.  When in "Significant Changes mode" it would transmit a new `_location` whenever the phone was plugged or unplugged from the charger.

The idea here is that this is a useful proxy for my intent to stay in that location for a while.  If I am at my home or work location and plugged in, I'm likely going to be there for a longer period, and not just driving by.  The first place I'm considering using this is in my home automation setup to trigger the "bedtime" event when everyone that is home is plugged in.

Thanks in advance, and keep up the great work!

**Comment by jpmens** [comment]
Thanks for your kind words and thank you for the idea. It is unlikely that this would be of great success if it's at all possible in the mobile OS' APIs. We'll keep this in mind but we'll not invest any effort at this point in time.

Obviously, this being an Open Source project, if you have or know of somebody with the capabilities to implement this feature we would entertain the thought of a pull request.

**Comment by m14t** [comment]
Thanks for the response!

I'm not an iOS developer, and so this is very much the results of hack-and-try, but this diff got the feature I was looking for working locally.

```
diff --git a/OwnTracks/OwnTracks/OwnTracking.m b/OwnTracks/OwnTracks/OwnTracking.m
index 6a8740d..4760355 100644
--- a/OwnTracks/OwnTracks/OwnTracking.m
+++ b/OwnTracks/OwnTracks/OwnTracking.m
@@ -413,6 +413,9 @@ static OwnTracking *theInstance = nil;
     if (batteryLevel >= 0) {
         [json setValue:@(batteryLevel) forKey:@"batt"];
     }
+    int state = [UIDevice currentDevice].batteryState;
+
+    [json setValue:@(state) forKey:@"battery_state"];
 
     return json;
 }
diff --git a/OwnTracks/OwnTracks/OwnTracksAppDelegate.m b/OwnTracks/OwnTracks/OwnTracksAppDelegate.m
index 74b2ea6..1196033 100644
--- a/OwnTracks/OwnTracks/OwnTracksAppDelegate.m
+++ b/OwnTracks/OwnTracks/OwnTracksAppDelegate.m
@@ -138,6 +138,12 @@ static const DDLogLevel ddLogLevel = DDLogLevelInfo;
                                                   usingBlock:^(NSNotification *note){
                                                       DDLogVerbose(@"UIApplicationUserDidTakeScreenshotNotification");
                                                   }];
+    [[NSNotificationCenter defaultCenter] addObserverForName:UIDeviceBatteryStateDidChangeNotification
+                                                      object:nil queue:nil usingBlock:^(NSNotification *note){
+                                                          CLLocation *location = [LocationManager sharedInstance].location;
+                                                          [self publishLocation:location trigger:@"bs"];
+                                                          [[GeoHashing sharedInstance] newLocation:location];
+                                                      }];
     return YES;
 }
 ```

A more formal version would likely want to do a mapping from Apple's constants to strings, incase android has other/different mappings.

Apple's docs: https://developer.apple.com/documentation/uikit/uidevicebatterystate

```
case unknown = 0
case unplugged = 1
case charging = 2
case full = 3
```

---

I want to be respectful of your project's scope and direction, and so if this is unwanted, I'll gladly leave this as it is here.

But if this looks like something you'd be interested in accepting, I'd be happy to create an pull-request for this and any documentation that should be updated.

Thanks again for this project and your time reading and considering this!

**Comment by ckrey** [comment]
(via @m14t)

Thanks for the response!

I'm not an iOS developer, and so this is very much the results of hack-and-try, but this diff got the feature I was looking for working locally.

diff --git a/OwnTracks/OwnTracks/OwnTracking.m b/OwnTracks/OwnTracks/OwnTracking.m
index 6a8740d..4760355 100644
--- a/OwnTracks/OwnTracks/OwnTracking.m
+++ b/OwnTracks/OwnTracks/OwnTracking.m
@@ -413,6 +413,9 @@ static OwnTracking *theInstance = nil;
     if (batteryLevel >= 0) {
         [json setValue:@(batteryLevel) forKey:@"batt"];
     }
+    int state = [UIDevice currentDevice].batteryState;
+
+    [json setValue:@(state) forKey:@"battery_state"];
 
     return json;
 }
diff --git a/OwnTracks/OwnTracks/OwnTracksAppDelegate.m b/OwnTracks/OwnTracks/OwnTracksAppDelegate.m
index 74b2ea6..1196033 100644
--- a/OwnTracks/OwnTracks/OwnTracksAppDelegate.m
+++ b/OwnTracks/OwnTracks/OwnTracksAppDelegate.m
@@ -138,6 +138,12 @@ static const DDLogLevel ddLogLevel = DDLogLevelInfo;
                                                   usingBlock:^(NSNotification *note){
                                                       DDLogVerbose(@"UIApplicationUserDidTakeScreenshotNotification");
                                                   }];
+    [[NSNotificationCenter defaultCenter] addObserverForName:UIDeviceBatteryStateDidChangeNotification
+                                                      object:nil queue:nil usingBlock:^(NSNotification *note){
+                                                          CLLocation *location = [LocationManager sharedInstance].location;
+                                                          [self publishLocation:location trigger:@"bs"];
+                                                          [[GeoHashing sharedInstance] newLocation:location];
+                                                      }];
     return YES;
 }
A more formal version would likely want to do a mapping from Apple's constants to strings, incase android has other/different mappings.

Apple's docs: https://developer.apple.com/documentation/uikit/uidevicebatterystate

case unknown = 0
case unplugged = 1
case charging = 2
case full = 3
I want to be respectful of your project's scope and direction, and so if this is unwanted, I'll gladly leave this as it is here.

But if this looks like something you'd be interested in accepting, I'd be happy to create an pull-request for this and any documentation that should be updated.

Thanks again for this project and your time reading and considering this!

**Comment by ckrey** [comment]
Dear @m14t, thank you for your input.

Your code shows how to include the battery state in the JSON payload. But this is not the critical point.

iOS does not allow an app to monitor battery (and a lot of other things) while in background. When OwnTracks "runs" in Significant Changes Mode, it is actually sleeping and is woken up by iOS when necessary (location changes, region enter/leave, and infrequently to do background fetches). This will not happen when battery status or battery level changes.

**Comment by m14t** [comment]
Thanks again for the reply.  You guys are great.

In my fever of getting this working, I was excited to share, but almost as soon as I posted this, I realized my haste and that I had only tested this while OwnTracks was the active application.  Once realizing this, I then understood @jpmens's earlier hesitation and deleted the comment, in hopes to not disturb you all.

Keep up the great work, and thank you for your patience with this iOS newb!
